brainiak/isfc.py

```diff
@@ -80,37 +86,50 @@ def isc(D, collapse_subj=True, return_p=False, num_perm=1000,
     n_vox = D.shape[0]
     n_subj = D.shape[2]
 
-    if return_p:
-        n_perm = num_perm
-    else:
-        n_perm = 0
-
-    ISC = np.zeros((n_vox, n_subj, n_perm + 1))
+    n_perm = num_perm*int(return_p)
+    max_null = np.zeros(n_perm, dtype=float_type)
+    min_null = np.zeros(n_perm, dtype=float_type)
+    ISC = np.zeros((n_vox, n_subj), dtype=float_type)
+
+    for loo_subj in range(n_subj):
+        group = np.mean(D[:, :, np.arange(n_subj) != loo_subj], axis=2)
+        subj = D[:, :, loo_subj]
+        for v in range(n_vox):
+            ISC[v, loo_subj] = stats.pearsonr(group[v, :],
+                                              subj[v, :])[0]
+    if collapse_subj:
+        ISC = np.mean(ISC, axis=1)
 
-    for p in range(n_perm + 1):
+    for p in range(n_perm):
+        # Randomize phases of D to create next null dataset
+        D = phase_randomize(D, random_state)
         # Loop across choice of leave-one-out subject
+        tmp_ISC = np.zeros((n_vox, n_subj), dtype=float_type)
         for loo_subj in range(n_subj):
             group = np.mean(D[:, :, np.arange(n_subj) != loo_subj], axis=2)
             subj = D[:, :, loo_subj]
             for v in range(n_vox):
-                ISC[v, loo_subj, p] = stats.pearsonr(group[v, :],
-                                                     subj[v, :])[0]
-
-        # Randomize phases of D to create next null dataset
-        D = phase_randomize(D, random_state)
-
-    if collapse_subj:
-        ISC = np.mean(ISC, axis=1)
+                tmp_ISC[v, loo_subj] = stats.pearsonr(group[v, :],
+                                                      subj[v, :])[0]
+        if not collapse_subj:
+            max_null[p] = np.max(tmp_ISC)
+            min_null[p] = np.min(tmp_ISC)
+        if collapse_subj:
+            tmp_ISC = np.mean(tmp_ISC, axis=1)
+            max_null[p] = np.max(tmp_ISC)
+            min_null[p] = np.min(tmp_ISC)

```

Comment:
We try not to repeat code. How about:
```
if collapse_subj:
    tmp_ISC = np.mean(tmp_ISC, axis=1)
max_null[p] = np.max(tmp_ISC)
min_null[p] = np.min(tmp_ISC)
```

Reply:
Thanks! Didn't notice that when I realized the max_null should always be a 1-D array. Fixed.